Update to wasmvm 2.0.0-rc.1

chipshort opened the pull request:
Very WIP, main goal is to get it to compile and pass current tests

TODOs:
- [x] supported capabilities are now `[]string`
- [x] nested Result changes
- [x] renaming
- [x] StoreCode gas
- [x] ValidateAddress
- [x] `Array[C]` type replacements
- [x] Adjust gas multiplier
- [x] TransferMsg.Memo
- [x] QueryRequest.Grpc
- [x] MsgResponses (see #1800)
- [x] IBCReceiveResponse.Acknowledgement == nil
- [x] Payload
- [x] Make `redactError` call conditional in `func (d MessageDispatcher) DispatchSubmessages`

chipshort commented on x/wasm/keeper/keeper.go:
```diff
@@ -169,7 +169,9 @@ func (k Keeper) create(ctx context.Context, creator sdk.AccAddress, wasmCode []b
 
 	sdkCtx.GasMeter().ConsumeGas(k.gasRegister.CompileCosts(len(wasmCode)), "Compiling wasm bytecode")
-	checksum, err = k.wasmVM.StoreCode(wasmCode)
+	gas := k.runtimeGasForContract(sdkCtx)
```
I think gasLeft explains the intent better here. I was just keeping the naming the same as in the other functions.

webmaster128 commented on x/wasm/keeper/api.go:
```diff
@@ -22,25 +24,32 @@ const (
 
+func validateAddress(human string) (uint64, error) {
+	_, err := sdk.AccAddressFromBech32(human)
+	return costValidate, err
```
This implementation does not yet check that the address is normalized since bech32 allows both uppercase and lower case inputs.

For a generic normalization check we need to convert it back and and check if we get the same result as the input.

chipshort commented:
@webmaster128 Any idea why the CI docker-image libwasmvm check is failing?
When I locally run `(cd cmd/wasmd && go run . query wasm libwasmvm-version)`, I get `2.0.0-rc.1`. Is the docker image an old version?

webmaster128 commented:
> @webmaster128 Any idea why the CI docker-image libwasmvm check is failing?
> When I locally run `(cd cmd/wasmd && go run . query wasm libwasmvm-version)`, I get `2.0.0-rc.1`. Is the docker image an old version?

I did not check the error yet but you need to update those lines for sure: https://github.com/CosmWasm/wasmd/blob/main/Dockerfile#L18-L21

webmaster128 commented:
I think it would be valuable to do as many of the _Probably in follow-up PRs:_ items in here as well. This PR then serves as a go-to-place for every other user of wasmvm (which right now is primarily wasmd forks and the Wasm Light Client).

pinosu commented:
Changes look good so far! To make the CI pass you need to update the[ Dockerfile](https://github.com/CosmWasm/wasmd/blob/main/Dockerfile#L18-L21) both version and checksum for libwasmvm.

chipshort commented:
@pinosu @webmaster128 Does one of you know the difference between the different `reflect.wasm` files and `reflect_1_1.wasm` used in tests? Because of the `Stargate` to `Any` rename, we either need to update them to the CW 2.0 version or we need to continue calling them with `stargate` json. Any opinions on that?

webmaster128 commented:
I think it would be pretty cool to test both, the old (stargate) and the new (any) message type to ensure wasmd supports both. I don't know about the 1_1 one. it was added here: https://github.com/CosmWasm/wasmd/pull/1055/files

webmaster128 commented:
Could you add the cosmwasm_2_0 capability to the README (on top of #1805)?

chipshort commented:
I rebased on main and added the capability to the readme.
I also fixed some more tests and added an updated reflect contract that is now used in most tests, as well as one test that uses both the old and new reflect contract for a `Stargate` / `Any` msg (`TestGroupWithContract` / `TestGroupWithNewReflectContract`). That should cover the backwards compatibility with the old `Stargate` variant.

To get the tests to pass, [this PR in wasmvm](https://github.com/CosmWasm/wasmvm/pull/519/files) is needed because the `Payload` is currently serialized incorrectly on the Go side. After that, the tests should pass and I can start working on the other new features.

chipshort commented:
Tests are passing 🎉

chipshort commented:
I added some code for the `redactError` change. Not sure if this is the best way to do it or if it would be better to have a completely separate error type for that to match against (like it is done for the `SystemError`)

pinosu commented on app/app.go:
```diff
@@ -646,7 +644,7 @@ func NewWasmApp(
 		wasmDir,
 		wasmConfig,
-		availableCapabilities,
+		AllCapabilities(),
```
👍

pinosu commented on app/sim_test.go:
```diff
@@ -331,6 +332,8 @@ func TestAppStateDeterminism(t *testing.T) {
 				logger = log.NewNopLogger()
 			}
 
+			appOptions.SetDefault(flags.FlagHome, t.TempDir()) // ensure a unique folder per run
```
nice! makes sense 👍

pinosu commented on Dockerfile:
```diff
@@ -15,13 +15,10 @@ RUN apk add git
-RUN sha256sum /lib/libwasmvm_muslc.x86_64.a | grep e660a38efb2930b34ee6f6b0bb12730adccb040b6ab701b8f82f34453a426ae7
-
-# Copy the library you want to the final location that will be found by the linker flag `-lwasmvm_muslc`
-RUN cp /lib/libwasmvm_muslc.${arch}.a /lib/libwasmvm_muslc.a
```
is this not needed anymore?

chipshort replied:
The new version of wasmvm now links against `wasmvm_muslc.x86_64` / `wasmvm_muslc.aarch64` instead of `wasmvm_muslc`, so the renaming is not needed anymore.
It was changed in this PR: https://github.com/CosmWasm/wasmvm/pull/514/files

webmaster128 replied:
Chris, could you add this change to https://github.com/CosmWasm/wasmvm/blob/main/docs/MIGRATING.md?

pinosu commented on tests/e2e/group_test.go:
```diff
@@ -74,3 +74,58 @@ func TestGroupWithContract(t *testing.T) {
+
+	var groupRsp group.MsgSubmitProposalResponse
+	require.NoError(t, chain.Codec.Unmarshal(execRsp.Data, &groupRsp))
+	// require.NotEmpty(t, groupRsp.ProposalId)
```
why this is commented? please remove if not needed

pinosu commented on tests/e2e/group_test.go:
```diff
@@ -74,3 +74,58 @@ func TestGroupWithContract(t *testing.T) {
 	assert.Equal(t, expBalanceAmount.String(), recipientBalance.String())
 }
+
+func TestGroupWithNewReflectContract(t *testing.T) {
```
Nice test!

pinosu commented on tests/e2e/reflect_helper.go:
```diff
@@ -16,9 +18,18 @@ import (
+// InstantiateStargateReflectContract stores and instantiates a pre 2.0 reflect contract instance.
+// This instance still expects the old CosmosMsg.Stargate variant instead of the new CosmosMsg.Any.
+func InstantiateStargateReflectContract(t *testing.T, chain *ibctesting.TestChain) sdk.AccAddress {
+	codeID := chain.StoreCodeFile("../../x/wasm/keeper/testdata/reflect.wasm").CodeID
```
Should we modify also the name of the file to give a meaningful name? Something like this
```suggestion
	codeID := chain.StoreCodeFile("../../x/wasm/keeper/testdata/stargate_reflect.wasm").CodeID
```

chipshort replied:
Funny, I was thinking something similar while implementing this. Will do!

pinosu commented on tests/e2e/reflect_helper.go:
```diff
@@ -37,12 +48,38 @@ type sdkMessageType interface {
+	// build the complete reflect message
+	reflectSendBz := []byte(fmt.Sprintf("{\"reflect_msg\":{\"msgs\":%s}}", vmMsgs))
+
+	// {"reflect_msg":{"msgs":[{"stargate":{"type_url":"mytype","value":"EjQ="}}]}}
```
remove

pinosu commented on x/wasm/keeper/api.go:
```diff
@@ -22,25 +26,39 @@ const (
 )
 
-func humanAddress(canon []byte) (string, uint64, error) {
+func humanizeAddress(canon []byte) (string, uint64, error) {
```
👍

pinosu commented on x/wasm/keeper/handler_plugin.go:
```diff
@@ -173,67 +175,69 @@ func NewIBCRawPacketHandler(ics4Wrapper types.ICS4Wrapper, channelKeeper types.C
-		return nil, nil, errorsmod.Wrap(err, "failed to marshal IBC send response")
+		return nil, nil, nil, errorsmod.Wrap(err, "failed to marshal IBC send response")
 	}
+	// TODO: fill msgResponse
```
is this pending?

chipshort replied:
Nice catch! Completely overlooked that

pinosu commented on x/wasm/keeper/handler_plugin_encoders_test.go:
```diff
@@ -421,13 +419,50 @@ func TestEncoding(t *testing.T) {
+		t.Run(name, func(t *testing.T) {
+			var ctx sdk.Context
+			encoder := DefaultEncoders(encodingConfig.Codec, wasmtesting.MockIBCTransferKeeper{})
+			res, err := encoder.Encode(ctx, tc.sender, "", tc.srcMsg)
```
probably the variable could be removed
```suggestion
			res, err := encoder.Encode(sdk.Context{}, tc.sender, "", tc.srcMsg)
```

pinosu commented on x/wasm/keeper/handler_plugin_encoders_test.go:
```diff
@@ -421,13 +419,50 @@ func TestEncoding(t *testing.T) {
+		addr1 = RandomAccountAddress(t)
+		addr2 = RandomAccountAddress(t)
+	)
+	valAddr := make(sdk.ValAddress, types.SDKAddrLen)
+	valAddr[0] = 12
+	valAddr2 := make(sdk.ValAddress, types.SDKAddrLen)
+	valAddr2[1] = 123
```
is this used anywhere?

chipshort replied:
No, leftover from copy-pasting. Removed

pinosu reviewed (changes requested):
Great work! 💯 
I did a first round review. I added some minor comments.
I will review it again before approving and merging it.

webmaster128 commented on README.md:
```diff
@@ -37,6 +37,7 @@ The following table shows which of the [latest capabilities](https://github.com/
 | cosmwasm_1_2 | x       | x       | x       |         |      |
 | cosmwasm_1_3 | x       | x       |         |         |      |
 | cosmwasm_1_4 | x       |         |         |         |      |
+| cosmwasm_2_0 |         |         |         |         |      |
```
Could you add a column that includes cosmwasm_2_0 for wasmd >= 0.51?

webmaster128 commented on tests/e2e/reflect_helper.go:
```diff
@@ -16,9 +18,18 @@ import (
+// InstantiateStargateReflectContract stores and instantiates a pre 2.0 reflect contract instance.
+// This instance still expects the old CosmosMsg.Stargate variant instead of the new CosmosMsg.Any.
+func InstantiateStargateReflectContract(t *testing.T, chain *ibctesting.TestChain) sdk.AccAddress {
+	codeID := chain.StoreCodeFile("../../x/wasm/keeper/testdata/stargate_reflect.wasm").CodeID
```
Woudn't it be easier to manage if we used the file name `reflect_1_4.wasm` here (and download file from https://github.com/CosmWasm/cosmwasm/releases/tag/v1.4.3). the file itself should not care too much why it is used here in the tests. No strong opinion on 1.2, 1.3, 1.4, or 1.5 – just something old.

webmaster128 commented on x/wasm/keeper/api.go:
```diff
@@ -14,6 +16,8 @@ const (
 	// DefaultGasCostCanonicalAddress is how much SDK gas we charge to convert to a canonical address format
 	DefaultGasCostCanonicalAddress = 4
+	// DefaultGasCostValidateAddress is how much SDK gas we charge to validate an address
+	DefaultGasCostValidateAddress = 9
```
```suggestion
	DefaultGasCostValidateAddress = DefaultGasCostHumanAddress + DefaultGasCostCanonicalAddress
```

does this work too?

webmaster128 commented on x/wasm/keeper/api.go:
```diff
@@ -22,25 +26,39 @@ const (
 var (
 	costHumanize            = DefaultGasCostHumanAddress * types.DefaultGasMultiplier
 	costCanonical           = DefaultGasCostCanonicalAddress * types.DefaultGasMultiplier
+	costValidate            = DefaultGasCostCanonicalAddress * types.DefaultGasMultiplier
```
```suggestion
	costValidate            = DefaultGasCostValidateAddress * types.DefaultGasMultiplier
```

chipshort replied:
Nice catch!

webmaster128 commented on x/wasm/keeper/api.go:
```diff
@@ -22,25 +26,39 @@ const (
+	if canonicalized.String() != human {
+		return costValidate, errors.New("address not normalized")
+	}
+	return costValidate, err
```
The end of the function is supposed to be the success case, right?

```suggestion
	return costValidate, nil
```

webmaster128 commented on x/wasm/keeper/api.go:
```diff
@@ -22,25 +26,39 @@ const (
 	return bz, costCanonical, err
 }
 
+func validateAddress(human string) (uint64, error) {
```
Do we have a place to write unit tests for those functions?

chipshort replied:
Currently not, but I can add an `api_test.go` and add tests there.

webmaster128 replied:
Yeah that would be great. At least for the new addition

chipshort replied:
I added a very basic test here.

webmaster128 reviewed (commented):
Good stuff! Not done reviwwing but here is a start

chipshort commented:
The test-system job seems to fail sporadically. Is it just a timing thing?

pinosu commented:
> The test-system job seems to fail sporadically. Is it just a timing thing?

Yes, unfortunately since we added the system test for the upgrade, sporadically it fails for timeout

webmaster128 commented:
If you look at the logs and the implementation of `AwaitBlockHeight` we see that we wait for block number 21, current block is 9 and block time is 1s with a timeout of 14s. Finalizing 12 one-second blocks within 14 seconds is very optimistic given that the blockTime parameter is used for `"--commit-timeout=" + s.blockTime.String(),`. But commit timeout is not the block time but a lower bound for the block time (https://docs.tendermint.com/v0.34/tendermint-core/configuration.html).  We use e.g.

```
			// Consensus settings
			config.Consensus.TimeoutPropose = 2000 * time.Millisecond
			config.Consensus.TimeoutProposeDelta = 500 * time.Millisecond
			config.Consensus.TimeoutPrecommitDelta = 500 * time.Millisecond
			config.Consensus.TimeoutPrevote = 1 * time.Second
			config.Consensus.TimeoutPrecommit = 1 * time.Second
			config.Consensus.TimeoutPrecommitDelta = 500 * time.Millisecond
			config.Consensus.TimeoutCommit = 1750 * time.Millisecond
```

to get block times of 2.11s in production or

```
# Custom settings for very fast blocks in CI
# We target 250ms block times with one validator.
sed -i"" \
  -e 's/^timeout_propose =.*$/timeout_propose = "100ms"/' \
  -e 's/^timeout_propose_delta =.*$/timeout_propose_delta = "100ms"/' \
  -e 's/^timeout_prevote =.*$/timeout_prevote = "100ms"/' \
  -e 's/^timeout_prevote_delta =.*$/timeout_prevote_delta = "100ms"/' \
  -e 's/^timeout_precommit =.*$/timeout_precommit = "100ms"/' \
  -e 's/^timeout_precommit_delta =.*$/timeout_precommit_delta = "100ms"/' \
  -e 's/^timeout_commit =.*$/timeout_commit = "230ms"/' \
  "config/config.toml"
```

to aim for 250ms block times in CI. So I think the `--commit-timeout` should be something like 90% of the desired block time.

webmaster128 commented on x/wasm/types/errors.go:
```diff
@@ -10,6 +10,10 @@ import (
 
+	// ContractErrorCodes are the error codes for errors returned by the contract
+	// Since contract execution is deterministic, the errors are also deterministic
+	ContractErrorCodes = []uint32{InstantiateErrorCode, ExecuteErrorCode, QueryErrorCode, MigrateErrorCode}
```
I think with this approach the chances are high that one of
- `ErrInstantiateFailed`
- `ErrExecuteFailed`
- `ErrQueryFailed`
- `ErrMigrationFailed`

is used in more than one place. E.g. `ErrExecuteFailed` is used many times throughout the codebase for vaious different things. `ErrMigrationFailed` is used in two places as well.

I think the safer approach is to explicitly mark errors as "Contract error" (as in the error case of

```go
// ContractResult is the raw response from the instantiate/execute/migrate calls.
// This is mirrors Rust's ContractResult<Response>.
type ContractResult struct {
	Ok  *Response `json:"ok,omitempty"`
	Err string    `json:"error,omitempty"`
}
```

in the places where we know for sure they are coming from the contract. Everything else by default remains redacted.

This would also play nicely with https://github.com/CosmWasm/wasmd/issues/1813 where we get migrations that don't even call into the contract, such that the idea of a "migration error" can be different things on two different levels.

webmaster128 replied:
The term "contract error" is also used here: https://github.com/CosmWasm/cosmwasm/pull/1646

chipshort replied:
That makes sense. So, we would add a new `ContractError` type and check against that? I was trying to do that at first, but I was not sure how to add the error codes and codespace with that.
After looking at it again, we could handle this in a similar way as the `WasmVMFlavouredError`.

webmaster128 replied:
> So, we would add a new ContractError type and check against that?

That was my first idea. However, I think I missed that the error must work well in the two scenarios:
1. created in a submessage and handled in reply
2. created in a transaction requiring code and codespace

> After looking at it again, we could handle this in a similar way as the WasmVMFlavouredError.

Interesting, I did not look into that before. However, I wonder if there is a way to attach some sort of "is deterministic" property to the normal Cosmos SDK error type that we can inspect in  `redactError` to keep it simpler.

chipshort replied:
I think that should be possible. I'll experiment a bit.

chipshort replied:
I added a proposal in 4e7e53422c1b49fcaf482c61524cba0d88abe625. It's a wrapper type around any error that marks it as deterministic for `redactError`. I only marked the errors coming from the contract as that for now (in theory most errors produced by wasmd should be deterministic as long as we don't change the message between versions.

webmaster128 replied:
Love it

> I only marked the errors coming from the contract as that for now (in theory most errors produced by wasmd should be deterministic as long as we don't change the message between versions).

We can add more and more errors over time which makes this solution extra powerful and elegant. Changing error text is usually not consensus breaking and happens in Cosmos SDK patch releases. But we can highlight that some error texts must be stable to get the extra benefit of having them non-redacted in the replies.

webmaster128 commented on x/wasm/keeper/msg_dispatcher.go:
```diff
@@ -189,6 +215,12 @@ func redactError(err error) error {
+	// Also do not redact any errors that are coming from the contract,
+	// as they are always deterministic
+	if codespace == types.DefaultCodespace && contains(types.ContractErrorCodes, code) {
+		return err
```
Should we unwrap what was wrapped before? Not sure if this is possible though

webmaster128 commented on x/wasm/types/errors.go:
```diff
@@ -164,3 +152,32 @@ func (e WasmVMFlavouredError) Wrap(desc string) error { return errorsmod.Wrap(e,
+}
+
+// Cause is the same as unwrap but used by ABCIInfo
+// By returning the wrapped error here, we ensure
```
This sentence seems broken

chipshort replied:
oops, I just stopped mid-sentence and forgot about it 😆

webmaster128 commented on x/wasm/types/errors.go:
```diff
@@ -164,3 +152,32 @@ func (e WasmVMFlavouredError) Wrap(desc string) error { return errorsmod.Wrap(e,
+	return e.error
+}
+
+// Cause is the same as unwrap but used by ABCIInfo
```
Is there an interface we can check this implementation against?

chipshort replied:
Unfortunately not. ABCIInfo has a [private interface](https://github.com/cosmos/cosmos-sdk/blob/35c4243da9c8c285a4b65117238dec9b9f7921c8/errors/errors.go#L300) for that and uses it to [iterate through the error chain](https://github.com/cosmos/cosmos-sdk/blob/35c4243da9c8c285a4b65117238dec9b9f7921c8/errors/abci.go#L92) until it finds an error with codespace / code.

webmaster128 replied:
okay, fine, thanks for checking

webmaster128 commented on x/wasm/keeper/msg_dispatcher.go:
```diff
@@ -209,18 +209,19 @@ func redactError(err error) error {
+	// We only check the top level error to avoid changes in the error chain becoming
+	// consensus-breaking.
+	if _, ok := err.(types.DeterministicError); ok {
+		return err
```
Should we unwrap here?

chipshort replied:
We could, but it doesn't make much of a difference since the DeterministicError has the same `Error()` message as its wrapped error.

webmaster128 replied:
I see, thanks

webmaster128 commented on x/wasm/types/errors_test.go:
```diff
@@ -94,3 +94,19 @@ func TestWasmVMFlavouredError(t *testing.T) {
+	err := MarkErrorDeterministic(inner)
+
+	// behaves like a wrapper around inner error
+	assert.Equal(t, inner.Error(), err.Error())
```
Could we add a sentence for this property in the `DeterministicError` docs? I think this is very powerful feature to have.

webmaster128 reviewed (commented):
Very very nice solution

chipshort commented:
> Do we have an integration test somewhere that shows how a submessage error is returned to a contract unredacted?

Good point, I added a test now that instantiates two reflect instances and makes one send a submessage that fails to the other one. It then checks the error message the first contract got.

pinosu commented on x/wasm/keeper/relay.go:
```diff
@@ -194,7 +200,11 @@ func (k Keeper) OnAckPacket(
 		return errorsmod.Wrap(types.ErrExecuteFailed, execErr.Error())
 	}
-	return k.handleIBCBasicContractResponse(ctx, contractAddr, contractInfo.IBCPortID, res)
+	if res.Err != "" {
```
should we add a check that `res != nil`?

chipshort replied:
Right now wasmvm only returns `nil` in case of an `err`, but it's probably better to add a check here anyways, just in case.
